### PR TITLE
Validate that locals have a corresponding `LocalDecl`

### LINKED ISSUE #73356: -Zvalidate-mir: check that all referenced locals exist
When encountering a `Local`, there must be a corresponding `LocalDecl` in the MIR body. If there isn't, later passes can panic due to indexing out of bounds of the `LocalDecl` array. The validation pass could detect this problem earlier.

### PR DESCRIPTION
**Pull request by camelid**
Fixes #73356.

### CODE REVIEW

**Comment by camelid** [comment]
How should I test that this works correctly?

**Comment by jonas-schievink** [comment]
> How should I test that this works correctly?

You can run `RUSTFLAGS_NOT_BOOTSTRAP="-Zvalidate-mir" ./x.py test`

**Comment by camelid** [comment]
Sorry, what I meant was how do I create something to test it on? This check will only fire when there's a bug in the compiler, so unless the idea is to mess with some other part of the compiler to introduce a bug, I'm not sure how to produce input to test it on.

**Comment by jonas-schievink** [comment]
Yeah, it's fine not to test this further.

File: compiler/rustc_mir/src/transform/validate.rs [hunk 1]
```diff
@@ -183,6 +183,13 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
 
 impl<'a, 'tcx> Visitor<'tcx> for TypeChecker<'a, 'tcx> {
     fn visit_local(&mut self, local: &Local, context: PlaceContext, location: Location) {
+        if !self.body.local_decls.contains(local) {
```
Comment (jonas-schievink) [inline compiler/rustc_mir/src/transform/validate.rs:186]:
```suggestion
        if self.body.local_decls.get(local).is_none() {
```

**Pull request merged**